molgenis-navigator/src/main/java/org/molgenis/navigator/copy/service/CopyServiceImpl.java

```diff
@@ -0,0 +1,121 @@
+import org.molgenis.navigator.util.ResourceCollector;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Transactional;
+
+@SuppressWarnings("unused")
+@Component
+public class CopyServiceImpl implements CopyService {
+
+  private final ResourceCollector resourceCollector;
+  private final MetaDataService metaDataService;
+  private final PackageCopier packageCopier;
+  private final EntityTypeCopier entityTypeCopier;
+
+  CopyServiceImpl(
+      ResourceCollector resourceCollector,
+      MetaDataService metaDataService,
+      PackageCopier packageCopier,
+      EntityTypeCopier entityTypeCopier) {
+    this.resourceCollector = requireNonNull(resourceCollector);
+    this.metaDataService = requireNonNull(metaDataService);
+    this.packageCopier = requireNonNull(packageCopier);
+    this.entityTypeCopier = requireNonNull(entityTypeCopier);
+  }
+
+  @Override
+  @Transactional(isolation = Isolation.SERIALIZABLE)
+  public Void copy(List<ResourceIdentifier> resources, String targetPackageId, Progress progress) {
+    try {
+      ResourceCollection resourceCollection = resourceCollector.get(resources);
+      Package targetPackage = getPackage(targetPackageId);
+      CopyState state = CopyState.create(targetPackage, progress);
+
+      copyResources(resourceCollection, state);
+
+    } catch (CodedRuntimeException exception) {
+      throw new CopyFailedException(exception);
+    }
+
+    return null;
+  }
+
+  private void copyResources(ResourceCollection resourceCollection, CopyState state) {
+    Progress progress = state.progress();
+    progress.setProgressMax(calculateMaxProgress(resourceCollection));
+    progress.progress(0, getMessage("progress-copy-started"));
+
+    copyPackages(resourceCollection.getPackages(), state);
+    copyEntityTypes(resourceCollection.getEntityTypes(), state);
+
+    progress.status(getMessage("progress-copy-success"));
+  }
+
+  private void copyPackages(List<Package> packages, CopyState state) {
+    if (!packages.isEmpty()) {
+      state.progress().status(getMessage("progress-copy-packages"));
+      packageCopier.copy(packages, state);
+    }
+  }
+
+  private void copyEntityTypes(List<EntityType> entityTypes, CopyState state) {
+    if (!entityTypes.isEmpty() || !state.entityTypesInPackages().isEmpty()) {
+      state.progress().status(getMessage("progress-copy-entity-types"));
+      entityTypeCopier.copy(entityTypes, state);
+    }
+  }
+
+  private Package getPackage(String targetPackageId) {
+    return targetPackageId != null
+        ? metaDataService
+            .getPackage(targetPackageId)
+            .orElseThrow(() -> new UnknownPackageException(targetPackageId))
+        : null;
+  }
+
+  @NotNull
+  private String getMessage(String key) {
+    return MessageSourceHolder.getMessageSource()
```

Comment:
How about using ```implements MessageSourceAware``` to avoid using the static MessageSourceHolder thus improving testability.

Reply:
See #7970